java-checks/src/main/java/org/sonar/java/checks/regex/DuplicatesInCharacterClassCheck.java

```diff
@@ -0,0 +1,113 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.regex;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.java.regex.RegexParseResult;
+import org.sonar.java.regex.ast.CharacterClassUnionTree;
+import org.sonar.java.regex.ast.CharacterRangeTree;
+import org.sonar.java.regex.ast.PlainCharacterTree;
+import org.sonar.java.regex.ast.RegexBaseVisitor;
+import org.sonar.java.regex.ast.RegexTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+
+@Rule(key = "S5869")
+public class DuplicatesInCharacterClassCheck extends AbstractRegexCheck {
+
+  private static final String MESSAGE = "Remove duplicates in this character class.";
+
+  @Override
+  public void checkRegex(RegexParseResult regexForLiterals, MethodInvocationTree mit) {
+    if (!regexForLiterals.hasSyntaxErrors()) {
+      DuplicateFinder duplicateFinder = new DuplicateFinder();
+      duplicateFinder.setActiveFlags(getFlags(mit));
+      duplicateFinder.visit(regexForLiterals.getResult());
+    }
+  }
+
+  private class DuplicateFinder extends RegexBaseVisitor {
+
+    @Override
+    public void visitCharacterClassUnion(CharacterClassUnionTree tree) {
+      List<RegexTree> duplicates = new ArrayList<>();
+      TreeMap<Character, Boolean> inCharacterClass = new TreeMap<>();
+      for (RegexTree element : tree.getCharacterClasses()) {
+        if (element.is(RegexTree.Kind.PLAIN_CHARACTER)) {
+          char ch = ((PlainCharacterTree) element).getCharacter();
+          processRange(duplicates, inCharacterClass, ch, ch, element);
+        } else if (element.is(RegexTree.Kind.CHARACTER_RANGE)) {
+          CharacterRangeTree range = (CharacterRangeTree) element;
+          char lower = range.getLowerBound().getCharacter();
+          char upper = range.getUpperBound().getCharacter();
+          processRange(duplicates, inCharacterClass, lower, upper, range);
+        }
+      }
+      if (!duplicates.isEmpty()) {
+        List<RegexIssueLocation> secondaries = duplicates.stream()
+          .skip(1)
+          .map(duplicate -> new RegexIssueLocation(duplicate, MESSAGE))
```

Comment:
Why not using another message here? it seems to me to be strange to have the same message for all secondaries.

Reply:
What would the other message be? Conceptually, the first duplicated character isn't any different than the others, so in principle we could even report one issue for each duplicated character. But of course that would produce way too much noise if a character class contains a lot of duplicates, so I think making the subsequent duplicates secondaries is preferable, but I don't know how their message should differ.